csharp/src/Ice/Reference.cs

```diff
@@ -1411,14 +1411,20 @@ private void ClearConnection(Connection connection)
                             return false;
                     }
 
-                    // If PreferNonSecure is false, filter out all non-secure endpoints
-                    return PreferNonSecure || endpoint.IsSecure;
+                    if (Protocol == Protocol.Ice1 && !PreferNonSecure)
+                    {
+                        return endpoint.IsAlwaysSecure;
+                    }
+
+                    // If we made it this far then the endpoint is either ice2 or it's ice1 and PreferNonSecure == true
+                    Debug.Assert(Protocol == Protocol.Ice2 || (Protocol == Protocol.Ice1 && PreferNonSecure));
+
+                    return true;
                 });
 
-                if (PreferNonSecure)
+                if (Protocol == Protocol.Ice1 && PreferNonSecure)
                 {
-                    // It's just a preference: we can fallback to secure endpoints.
-                    filteredEndpoints = filteredEndpoints.OrderBy(endpoint => endpoint.IsSecure);
+                    filteredEndpoints = filteredEndpoints.OrderBy(endpoint => !endpoint.IsAlwaysSecure);
                 }
```

Comment:
PreferNonSecure applies to all protocols. With ice1, it means try first non-secure endpoints, since an ice1 endpoint is always secure or always non-secure.

Reply:
> PreferNonSecure applies to all protocols. With ice1, it means try first non-secure endpoints, since an ice1 endpoint is always secure or always non-secure.

Ok so with Ice2, you might get a secure connection even if `prefereNonSecure = true`? 

How can the proxy ensure it will use a non-secure connection if it wants for example to transmit non-sensitive data and doesn't want to pay the additional cost of TLS?

Reply:
It cannot: there is a transparent fallback to TLS. And that's why it's called _prefer_ non-secure.